Issue 15828 - Notification comment support in the new comment editor

RenanLukas opened the pull request:
Fixes #15828

Currently only Site Comments are supported in the new comment editor. The goal of this PR is to also allow Notification Comments to be shown and editable in the new comment editor (instead of the legacy editor). It also tries to make it easier to support other Comment types in the future (e.g. Reader).

<p float="left">
<img width="400" alt="Screen Shot 2022-02-03 at 2 43 47 PM" src="https://user-images.githubusercontent.com/14964993/152401230-c3a2bd88-d59a-41e4-b6af-b2db0f132ce7.png">
<img width="400" alt="Screen Shot 2022-02-03 at 2 43 52 PM" src="https://user-images.githubusercontent.com/14964993/152401253-4f2f9531-cb72-4cc1-9f00-468534db8398.png">
</p>

<p float="left">
<img width="400" alt="Screen Shot 2022-02-03 at 2 43 59 PM" src="https://user-images.githubusercontent.com/14964993/152401259-85a8e2e6-6540-46a8-a648-dcecc74ea6e5.png">
<img width="400" alt="Screen Shot 2022-02-03 at 2 44 02 PM" src="https://user-images.githubusercontent.com/14964993/152401266-b4c16820-5368-4209-82c2-2376d48b9761.png">
</p>

To test:
1 - Login with an user that has at least one notification of a comment;
2 - Select any site;
3 - Tap on "Notifications" bottom navigation icon to open Notifications screen;
4 - Tap on "Comments" tab inside Notifications screen;
5 - Tap on the comment you want to test;
6 - Tap on "More" button;
7 - Tap on "Edit" item. The notification comment edit should be opened in the new comment editor;
8 - Edit the comment text;
9 - Tap on "Done";
10 - The comment text should be updated.

## Regression Notes
1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
-

3. What automated tests I added (or what prevented me from doing so)
- `CommentEssentialsTest.kt`
- `CommentSourceTest.kt`
- `CommentEssentialsExtensionKtTest.kt`
- `GetCommentUseCaseTest.kt`
- Included more tests in `UnifiedCommentsEditViewModelTest.kt`

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

RenanLukas committed:
Extract CommentSource class

RenanLukas committed:
Migrate CommentSource to Kotlin

RenanLukas committed:
Implement unit tests for CommentSource

RenanLukas committed:
Refactor CommentDetailFragment Intent creation

RenanLukas committed:
Open UnifiedCommentsEditActivity instead of EditCommentActivity for all CommentSource types (if feature flag is enabled)

RenanLukas committed:
Add CommentIdentifier class (as implemented in @khaykov branch)

RenanLukas committed:
Integrate CommentIdentifier with UnifiedCommentsEdit flow

RenanLukas committed:
Implement NotificationsTableWrapper

RenanLukas committed:
Update NotificationCommentIdenfitier ID type

RenanLukas committed:
Support notification comment in new editor: Use CommentIdentifier to decide where to get the comment from

RenanLukas committed:
Fix UnifiedCommentsEditViewModel unit tests

RenanLukas committed:
Merge branch 'trunk' into issue/15828-migrate-edited-comment-notification-logic-to-new-editor

RenanLukas committed:
Extract and implement unit tests for CommentEssentials

RenanLukas committed:
Add new unit tests for UnifiedCommentsEditViewModel

RenanLukas committed:
Fix detekt, ktlint and checkstyle

RenanLukas committed:
Implement save edited notification comment in new editor

RenanLukas committed:
Rename downloadNoteAndUpdateDB parameter

RenanLukas committed:
Update UnifiedCommentsEditViewModel unit tests considering saving of edited notification comment

RenanLukas committed:
Fix detekt warning for UnifiedCommentsEditViewModel: TooManyFuncions. Logic can be split among UseCases later.

RenanLukas committed:
Fix unresolved reference in UnifiedCommentsEditViewModelTest

RenanLukas committed:
Fix edit notification comment in new editor not fetching web address and email address

RenanLukas committed:
Update UnifiedCommentsEditViewModel unit tests

RenanLukas committed:
Fix detekt: MaxLineLength in UnifiedCommentsEditViewModel

RenanLukas committed:
Fetch remote notification comment if not found locally in UnifiedCommentsEditViewModel

RenanLukas committed:
Move remoteCommentId property to CommentIdentifier sealed class

RenanLukas committed:
Extract comment get logic to GetCommentUseCase

RenanLukas committed:
Extract CommentEssentials extension isNotEqualTo to separate file and add unit tests

RenanLukas committed:
Update UnifiedCommentsEditViewModel unit tests

RenanLukas committed:
Fix checkstyle of CommentEssentialsExtensionKtTest

RenanLukas committed:
Update RELEASE-NOTES with notification comments support in the new editor (19.3)

RenanLukas committed:
Merge branch 'trunk' into issue/15828-migrate-edited-comment-notification-logic-to-new-editor

RenanLukas committed:
Remove extra line in RELEASE-NOTES

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/comments/unified/UnifiedCommentsEditViewModel.kt:
```diff
@@ -206,49 +183,85 @@ class UnifiedCommentsEditViewModel @Inject constructor(
+        } ?: showUpdateCommentError()
+    }
+
+    private suspend fun updateCommentEntityLocalDatabase(
```
About the method name - `updateEditComment` pushes the comment to sever as well, but I don't think that VM needs to know about what is happening there :)

RenanLukas replied:
Good catch, thank you! I had this decision in the VM before extracting it to the `UseCase` and completely missed this name 😅 
Changing it

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/notifications/utils/NotificationsActions.java:
```diff
@@ -80,7 +80,7 @@ public void onErrorResponse(VolleyError error) {
     }
 
-    public static void downloadNoteAndUpdateDB(final String noteID, final RestRequest.Listener respoListener,
+    public static void downloadNoteAndUpdateDB(final String noteID, final RestRequest.Listener requestListener,
```
Good catch 👍

RenanLukas committed:
Use -1 instead of 0 for CommentEssentials default ID and update unit tests

khaykov commented on WordPress/src/test/java/org/wordpress/android/ui/comments/unified/usecase/GetCommentUseCaseTest.kt:
```diff
@@ -0,0 +1,110 @@
+        whenever(commentsStore.getCommentByLocalSiteAndRemoteId(siteModel.id, remoteCommentId))
+                .thenReturn(emptyList())
+        whenever(commentsStore.fetchComment(siteModel, remoteCommentId, null))
+                .thenReturn(CommentsActionPayload(CommentsActionData(emptyList(), 1)))
```
NP: Doesn't affect the test, but the `rowsAffected` in `CommentsActionData` will probably be 0 in this case :)

khaykov commented on WordPress/src/test/java/org/wordpress/android/ui/comments/viewmodels/UnifiedCommentsEditViewModelTest.kt:
```diff
@@ -48,97 +54,153 @@ class UnifiedCommentsEditViewModelTest : BaseUnitTest() {
+    }
+
+    @Test
+    fun `Should map CommentIdentifier to default CommentEssentials if CommentIdentifier comment not found`() {
```
NP: Wonder if this method signature should have `= test {...}` instead of `{ test{...} }`

khaykov commented on WordPress/src/test/java/org/wordpress/android/ui/comments/viewmodels/UnifiedCommentsEditViewModelTest.kt:
```diff
@@ -48,97 +54,153 @@ class UnifiedCommentsEditViewModelTest : BaseUnitTest() {
+    }
+
+    @Test
+    fun `Should map CommentIdentifier to default CommentEssentials if CommentIdentifier not handled`() {
```
Do we want the test to use `runBlocking` (`= test{...}`) or is it ok like this?

khaykov reviewed (approved):
Thanks for the PR @RenanLukas ! Looks good and works well :) I left couple of small nitpicks in the code :)

As we discussed, let's keep this PR unmerged, until we can update the content of the comment in notification details (not list) after the edit.

RenanLukas committed:
Implement code review improvements

RenanLukas commented:
Thank you for the review @khaykov 🎉  I've addressed the comments. 

> As we discussed, let's keep this PR unmerged, until we can update the content of the comment in notification details (not list) after the edit.

Deal 👍 I'm on it

RenanLukas committed:
Add callbacks to update edited notification comment edited in comment details screen

RenanLukas committed:
Split CommentUserNoteBlock configureView method

RenanLukas committed:
Update notification comment content in comment details after editing

RenanLukas committed:
Fix checkstyle errors

RenanLukas committed:
Implement VM logic to disable edit name, url and email for notification comment in new editor

RenanLukas committed:
Add check to avoid NPE when calling refreshBlocksForEditedComment in CommentDetailFragment

RenanLukas committed:
Apply input enabled settings in UnifiedCommentsEditFragment based on the CommentIdentifier

RenanLukas committed:
Update unit tests for UnifiedCommentsEditViewModelTest considering input enabled settings based on comment identifier

RenanLukas committed:
Update notification entity when comment notification edited in the new editor

RenanLukas committed:
Refresh note blocks in comment details screen after notification comment edited

RenanLukas commented:
This PR depends on https://github.com/wordpress-mobile/WordPress-Android/pull/15918

RenanLukas committed:
Remove Nullable annotations not found by CI

khaykov committed:
Merge pull request #15918 from wordpress-mobile/issue/15910-fix-notification-comment-details-content-not-updated-after-editing

Issue 15910 - Fix notification comment details content not updated after editing

RenanLukas committed:
Merge branch 'trunk' into issue/15828-migrate-edited-comment-notification-logic-to-new-editor

khaykov reviewed (approved):
Thanks for addressing the issues from the review, and fixing the issue with notification content not updating. Now everything works as expected :)
